### PR TITLE
branchandbound algo

### PR DESCRIPTION
**Pull request by Aadesha**
I will add tests soon.

### CODE REVIEW

File: src/branchandbound.jl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+end
+
+function _branchandbound(p::Union{TaylorN{T},Taylor1{T}},
+                       dom::Union{IntervalBox{N,W},Interval{W}}, ϵ::Number) where {N,T,W}
+    K = 1
+    Rperv = evaluate(p, dom)
+    D1, D2 = bisect(dom)
+    D = [D1, D2]
+    R = [evaluate(p, D[i]) for i = 1:length(D)]
+    Rnext = Interval(minimum(R[i].lo for i = 1:length(R)),
```
Comment (mforets) [inline src/branchandbound.jl]: `Rnext = Interval(minimum(inf.(R)), maximum(sup.(R)))` (same in other occurrences)

File: src/branchandbound.jl [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            D, R = devide_dom!(p, D, R, min_index)
+            Rnext = Interval(minimum(R[i].lo for i=1:length(R)),
+                             maximum(R[i].hi for i=1:length(R)))
+        end
+    end
+    return Interval(minimum(R[i].lo for i=1:length(R)),
+                    maximum(R[i].hi for i=1:length(R)))
+end
+
+function devide_dom!(p::Union{TaylorN{T},Taylor1{T}},
```
Comment (mforets) [inline src/branchandbound.jl]: `devide_dom!` -> `divide_dom!`

File: test/univariate.jl [hunk 3]
```diff
@@ -13,8 +13,11 @@
         elseif solver == :SumOfSquares
             # solver returns negative -1.06259e-06
             @test abs(inf(r)) ≤ 1e-5 && sup(r) ≤ 1e-5
+        elseif solver == :branchandbound
```
Comment (mforets) [inline test/univariate.jl]: `:branchandbound` -> `:BranchandBound`?

File: src/branchandbound.jl [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+end
+
+function branchandbound(p::TaylorN{T}, dom::IntervalBox{N,W}, ϵ::Number) where{T,N,W}
+    return _branchandbound(p, dom, ϵ)
+end
+
+function _branchandbound(p::Union{TaylorN{T},Taylor1{T}},
+                       dom::Union{IntervalBox{N,W},Interval{W}}, ϵ::Number) where {N,T,W}
+    K = 1
+    Rperv = evaluate(p, dom)
```
Comment (mforets) [inline src/branchandbound.jl]: what is particular of `p::Union{TaylorN{T},Taylor1{T}}` here? in other words, can you let `f::Function` and let (*) `evaluate(p, dom) -> enclose_TaylorModels(f, dom)`? 

(*) as a first step; later on we'd like to pass the options and more generally to use a custom solver

File: src/branchandbound.jl (hunk 1 above)
Comment (Aadesha) [reply src/branchandbound.jl]: done

File: src/branchandbound.jl (hunk 2 above)
Comment (Aadesha) [reply src/branchandbound.jl]: done

File: test/univariate.jl (hunk 3 above)
Comment (Aadesha) [reply test/univariate.jl]: done

File: src/branchandbound.jl (hunk 4 above)
Comment (Aadesha) [reply src/branchandbound.jl]: I will try it locally first

File: test/univariate.jl [hunk 5]
```diff
@@ -13,6 +13,9 @@
         elseif solver == :SumOfSquares
             # solver returns negative -1.06259e-06
             @test abs(inf(r)) ≤ 1e-5 && sup(r) ≤ 1e-5
+        elseif solver == :BranchandBound
```
Comment (mforets) [inline test/univariate.jl]: remove this case? if the solver returns a relative precision of `[1.21721e-15, 0]` there is no need to special case it since the bounds nonnegative (and way smaller than 1e-5)

File: src/branchandbound.jl [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    D1, D2 = bisect(dom)
+    D = [D1, D2]
+    R = [evaluate(p, D[i]) for i = 1:length(D)]
+    Rnext = _Rnext(R)
+    while  (Rperv.hi - Rnext.hi) <= ϵ*(Rnext.hi - Rnext.lo) &&
+           (Rperv.lo - Rnext.lo) <= ϵ*(Rnext.hi - Rnext.lo) && (K <= 1000)
```
Comment (mforets) [inline src/branchandbound.jl]: there is something odd here -- the same condition seems to be checked twice.

File: src/branchandbound.jl [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+function _branchandbound(p::Union{TaylorN{T},Taylor1{T}},
+                       dom::Union{IntervalBox{N,W},Interval{W}}, ϵ::Number) where {N,T,W}
+    K = 1
+    Rperv = evaluate(p, dom)
+    D1, D2 = bisect(dom)
+    D = [D1, D2]
+    R = [evaluate(p, D[i]) for i = 1:length(D)]
+    Rnext = _Rnext(R)
+    while  (Rperv.hi - Rnext.hi) <= ϵ*(Rnext.hi - Rnext.lo) &&
```
Comment (mforets) [inline src/branchandbound.jl]: `Rnext.hi - Rnext.lo` -> `diam(Rnext)`

File: src/branchandbound.jl (hunk 4 above)
Comment (mforets) [reply src/branchandbound.jl]: what does `Rperv` stand for? can you use more descriptive variable names?

Comment (Aadesha) [reply src/branchandbound.jl]: Bound over Range achieved by dividing domain on previous iteration

File: src/branchandbound.jl (hunk 6 above)
Comment (Aadesha) [reply src/branchandbound.jl]: Ok:)

File: test/univariate.jl (hunk 5 above)
Comment (Aadesha) [reply test/univariate.jl]: Yes Indeed.

File: src/branchandbound.jl (hunk 4 above)
Comment (mforets) [reply src/branchandbound.jl]: ok so can you call it, for example, `Rprevious` or `Rprev`?

Comment (Aadesha) [reply src/branchandbound.jl]: Yes

File: src/branchandbound.jl [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    D[index] = D1
+    DD = push!(D[1:index], D2)
+    D = append!(DD, D[(index+1):length(D)])
+    R[index] = evaluate(p, D[index])
+    RR = append!(R[1:index], evaluate(p,D[index+1]))
+    R = append!(RR, R[(index+1):length(R)])
+    return D, R
+end
+
+function _Rnext(R::Array{Interval{T}}) where{T}
```
Comment (mforets) [inline src/branchandbound.jl]: simpler: 

```julia
@inline _Rnext(R::Vector{Interval}) = Interval(minimum(inf.(R)), maximum(sup.(R)))
```
(note that `Array` is a possibly multidimensional array; either use `Array{..., 1}` or `Vector{...}` which is a convenience alias):

File: src/branchandbound.jl [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+function branchandbound(p::TaylorN{T}, dom::IntervalBox{N,W}, ϵ::Number) where{T,N,W}
+    return _branchandbound(p, dom, ϵ)
+end
+
+function _branchandbound(p::Union{TaylorN{T},Taylor1{T}},
+                       dom::Union{IntervalBox{N,W},Interval{W}}, ϵ::Number) where {N,T,W}
+    K = 1
+    Rperv = evaluate(p, dom)
+    D1, D2 = bisect(dom)
+    D = [D1, D2]
```
Comment (mforets) [inline src/branchandbound.jl]: why not use `D = bisect(dom)` ?

File: src/branchandbound.jl [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    R = [evaluate(p, D[i]) for i = 1:length(D)]
+    Rnext = _Rnext(R)
+    while  (Rperv.hi - Rnext.hi) <= ϵ*diam(Rnext) &&
```
Comment (mforets) [inline src/branchandbound.jl]: These conditions look odd, since `(Rperv.lo - Rnext.lo) <= ϵ*diam(Rnext)` always holds (the lhs is negative or zero); i would expect to stay refining as long as the improvement in the iteration is "big", like so:

```julia
while (Rprev.hi - Rnext.hi) >= ϵ*diam(Rnext) && (Rnext.lo - Rprev.lo) >= ϵ*diam(Rnext) && (K <= Kmax)
...
```

Otherwise, please add a docstring explaining the algorithm.

File: src/branchandbound.jl [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+           (Rperv.lo - Rnext.lo) <= ϵ*diam(Rnext) && (K <= 1000)
```
Comment (mforets) [inline src/branchandbound.jl]: the constant `K = 1000` here seems rather arbitrary, i would suggest that you set `const Kmax = 1000` then check if `K` is bigger than this value raise an error that convergence was not achieved.

File: src/branchandbound.jl (hunk 9 above)
Comment (Aadesha) [reply src/branchandbound.jl]: Then D will be Tuple `Tuple{IntervalBox{2,Float64},IntervalBox{2,Float64}}`  But D should be `2-element Array{IntervalBox{2,Float64},1}`

File: src/branchandbound.jl (hunk 10 above)
Comment (Aadesha) [reply src/branchandbound.jl]: 1> Yes doctoring would explain if we keep no of iteration(K) as input but I would try with        some examples if we could remove the `Kmax` altogether 
2> Thanks what you suggest is correct and I will make changes soon

File: src/branchandbound.jl (hunk 11 above)
Comment (Aadesha) [reply src/branchandbound.jl]: answered above

File: src/branchandbound.jl (hunk 7 above)
Comment (Aadesha) [reply src/branchandbound.jl]: done

File: src/branchandbound.jl (hunk 8 above)
Comment (Aadesha) [reply src/branchandbound.jl]: done

File: src/branchandbound.jl (hunk 9 above)
Comment (mforets) [reply src/branchandbound.jl]: true, but this doesn't answer my question -- where is the restriction to use a vector instead of a tuple?

File: src/branchandbound.jl (hunk 8 above)
Comment (mforets) [reply src/branchandbound.jl]: sorry, but it does not seem to be done -- see the online diff at https://github.com/JuliaReach/RangeEnclosures.jl/pull/7/

File: src/branchandbound.jl (hunk 9 above)
Comment (Aadesha) [reply src/branchandbound.jl]: Yes, but vector is also fine, what do you think?

File: src/branchandbound.jl (hunk 8 above)
Comment (Aadesha) [reply src/branchandbound.jl]: 1> I have Pushed in my forked file by mistake 
2> It should be 
```julia
@inline _Rnext(R::Vector{Interval{T},1}) where{T} = Interval(minimum(inf.(R)), maximum(sup.(R)))
```

File: src/branchandbound.jl (hunk 9 above)
Comment (mforets) [reply src/branchandbound.jl]: if there is no reason to allocate a new vector, i would not allocate a new vector :)

File: src/branchandbound.jl [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+end
+
+function branchandbound(p::TaylorN{T}, dom::IntervalBox{N,W}, ϵ::Number) where{T,N,W}
+    return _branchandbound(p, dom, ϵ)
+end
+
+function _branchandbound(p::Union{TaylorN{T},Taylor1{T}},
+                       dom::Union{IntervalBox{N,W},Interval{W}}, ϵ::Number) where {N,T,W}
+    K = 1
+    Rperv = evaluate(p, dom) #Rprevious
```
Comment (mforets) [inline src/branchandbound.jl]: sorry, but i still read `Rperv` here. it should be `Rprev`.

File: src/branchandbound.jl [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+function branchandbound(p::Taylor1{T}, dom::Interval{W}, ϵ::Number) where{T,W}
+    return _branchandbound(p, dom, ϵ)
+end
+
+function branchandbound(p::TaylorN{T}, dom::IntervalBox{N,W}, ϵ::Number) where{T,N,W}
+    return _branchandbound(p, dom, ϵ)
+end
+
+function _branchandbound(p::Union{TaylorN{T},Taylor1{T}},
+                       dom::Union{IntervalBox{N,W},Interval{W}}, ϵ::Number) where {N,T,W}
```
Comment (mforets) [inline src/branchandbound.jl]: please read the JuliaReach coding guidelines and apply them, esp. regarding the use of whitespace: https://juliareach.github.io/JuliaReachDevDocs/latest/guidelines/#Source-code-guidelines-1

Comment (Aadesha) [reply src/branchandbound.jl]: 1> Yes, I will be adding those things soon

File: .travis.yml [hunk 14]
```diff
@@ -19,5 +19,5 @@ jobs:
       os: linux
       script:
         - julia --project=docs/ -e 'using Pkg; Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate();'
-        - julia --project=docs/ docs/make.jl
+        - travis_wait 20 julia --project=docs/ docs/make.jl
```
Comment (mforets) [inline .travis.yml]: this change is unrelated to your PR; please rebase with master

File: src/branchandbound.jl [hunk 15]
```diff
@@ -0,0 +1,80 @@
+using TaylorModels
```
Comment (mforets) [inline src/branchandbound.jl]: it is not needed to do `using TaylorModels` as it has been already loaded in `taylormodels.jl`

Comment (Aadesha) [reply src/branchandbound.jl]: done

File: .travis.yml (hunk 14 above)
Comment (Aadesha) [reply .travis.yml]: done

File: .travis.yml [hunk 16]
```diff
@@ -19,5 +19,5 @@ jobs:
       os: linux
       script:
         - julia --project=docs/ -e 'using Pkg; Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate();'
-        - julia --project=docs/ docs/make.jl
+        - Julia --project=docs/ docs/make.jl
```
Comment (mforets) [inline .travis.yml]:
```suggestion
        - julia --project=docs/ docs/make.jl
```
